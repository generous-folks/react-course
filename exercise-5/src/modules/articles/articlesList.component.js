import React, { useEffect, useState } from 'react';

import Grid from '@material-ui/core/Grid';

import { ArticleCard } from './articleCard.component';
import { getArticles } from '../../utils/api.utils';

export function ArticlesList() {
  const [articles, setArticles] = useState([]);

  useEffect(() => {
    if (articles.length !== 0) {
      return;
    }
    getArticles()
      .then(setArticles)
      .catch(console.error);
  }, [articles]);

  return (
    <Grid container spacing={4}>
      {articles.map(article => (
        <ArticleCard key={article.id} article={article} />
      ))}
    </Grid>
  );
}
