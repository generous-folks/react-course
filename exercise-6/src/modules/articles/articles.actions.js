import { getArticles } from '../../utils/api.utils';

export const RECEIVED_ARTICLES = 'articles/RECEIVED_ARTICLES';

export const requestArticles = () => async dispatch => {
  const articles = await getArticles();

  return dispatch({ type: RECEIVED_ARTICLES, articles });
};
