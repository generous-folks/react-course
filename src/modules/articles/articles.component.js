import React from 'react'

const CartStateContext = React.createContext()
const CartDispatchContext = React.createContext()

function cartReducer(state, action) {
  switch (action.type) {
    case 'increment': {
      return { cart: state.cart + 1 }
    }
    case 'decrement': {
      return { cart: state.cart - 1 }
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}

function CartProvider({ children }) {
  const [state, dispatch] = React.useReducer(cartReducer, { cart: 0 })
  return (
    <CartStateContext.Provider value={state}>
      <CartDispatchContext.Provider value={dispatch}>
        {children}
      </CartDispatchContext.Provider>
    </CartStateContext.Provider>
  )
}

function useCartState() {
  const context = React.useContext(CartStateContext)
  if (context === undefined) {
    throw new Error('useCartState must be used within a CartProvider')
  }
  return context
}

function useCartDispatch() {
  const context = React.useContext(CartDispatchContext)
  if (context === undefined) {
    throw new Error('useCartDispatch must be used within a CartProvider')
  }
  return context
}

export { CartProvider, useCartState, useCartDispatch }
