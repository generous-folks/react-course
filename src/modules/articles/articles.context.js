import React from 'react'

const CartStateContext = React.createContext()
const CartDispatchContext = React.createContext()


function useCartState() {
  const context = React.useContext(CartStateContext)
  if (context === undefined) {
    throw new Error('useCartState must be used within a CartProvider')
  }
  return context
}

function useCartDispatch() {
  const context = React.useContext(CartDispatchContext)
  if (context === undefined) {
    throw new Error('useCartDispatch must be used within a CartProvider')
  }
  return context
}

function useCart() {
  return [useCartState(), useCartDispatch()];
}

export { useCart, useCartState, useCartDispatch }
