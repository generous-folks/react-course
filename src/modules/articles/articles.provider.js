
import React from 'react'
import PropTypes from 'prop-types'
import { cartReducer, initialState } from './cart.reducer'
import { CartStateContext, CartDispatchContext } from './cart.context'


export const CartProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(cartReducer, initialState)
  return (
    <CartStateContext.Provider value={state}>
      <CartDispatchContext.Provider value={dispatch}>
        {children}
      </CartDispatchContext.Provider>
    </CartStateContext.Provider>
  )
}

CartProvider.propTypes = {
  children: PropTypes.element,
}
