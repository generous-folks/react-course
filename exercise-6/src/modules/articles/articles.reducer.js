import { RECEIVED_ARTICLES } from './articles.actions';

export const initialState = {
  articles: [],
};

export const articlesReducer = (state, action) => {
  switch (action.type) {
    case RECEIVED_ARTICLES: {
      return { ...state, articles: [...state.articles, ...action.articles] };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};
