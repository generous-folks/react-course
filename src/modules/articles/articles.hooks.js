import { useEffect } from 'react'
import { useArticles } from "./articles.context";
import { requestArticles } from './articles.actions';

export const useArticlesSelector = () => {
  const [{ articles }, dispatch] = useArticles();

  useEffect(() => {
    if(articles.length === 0) {
      dispatch(requestArticles());
    }
  }, [dispatch, articles])

  return articles;
}
