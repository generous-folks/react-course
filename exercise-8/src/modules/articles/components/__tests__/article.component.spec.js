import React from 'react';

import { shallow } from 'enzyme';

import { Article } from '../article.component';
import MockedArticles from '../../../../../../fixtures/articles.json';
import * as selector from '../../articles.selectors';

jest.mock('../../articles.selectors');
selector.useArticlesSelector = jest.fn().mockReturnValue(MockedArticles);

const getWrapper = id => shallow(<Article id={id} />);

beforeEach(() => {
  jest.clearAllMocks();
});

describe('<Article />', () => {
  describe('Snapshot', () => {
    it('should render correctly', () => {
      expect(getWrapper(MockedArticles[0].slug)).toMatchSnapshot();
    });
  });

  describe('Markup checks', () => {
    it('should contain the correct markup', () => {
      expect(
        getWrapper(MockedArticles[0].slug)
          .find('ArticleCard')
          .exists(),
      ).toBeTruthy();
    });

    it('should be null when no id is passed', () => {
      expect(getWrapper('foo')).toBeEmptyRender();
    });
  });
});
