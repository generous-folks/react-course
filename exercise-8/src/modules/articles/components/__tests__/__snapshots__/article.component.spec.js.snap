// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> Snapshot should render correctly 1`] = `
<ArticleCard
  article={
    Object {
      "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam scelerisque nunc sagittis sollicitudin placerat. Maecenas in consectetur magna, egestas auctor massa. Aenean velit nunc, pulvinar id commodo in, ultricies nec velit. Integer efficitur porttitor feugiat. Nulla facilisi. Nulla sit amet mauris eu metus pellentesque faucibus. Nunc mattis dapibus urna at vulputate. Integer lobortis nisl quis arcu sodales sodales. Praesent malesuada et metus non vestibulum. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Nullam diam orci, ornare ut aliquet et, porta eget arcu.",
      "id": "O4GVOHgdvjkdivlRxQPlZpV",
      "image": "https://picsum.photos/602/800",
      "name": "Massive Oak Table",
      "price": 1000,
      "slug": "massive-oak-table-1975",
      "type": "furniture",
      "year": "1975",
    }
  }
/>
`;
